src/NetscriptFunctions/Infiltration.ts

```diff
@@ -81,5 +112,59 @@ export function NetscriptInfiltration(): InternalAPI<NetscriptInfiltation> {
       const locationName = getEnumHelper("LocationName").nsGetMember(ctx, _locationName);
       return calculateInfiltrationData(ctx, locationName);
     },
+    getMarketDemand: (ctx) => () => {
+      checkAccess(ctx);
+      return calculateMarketDemandMultiplier(Date.now(), false);
+    },
+    startInfiltration: (ctx) => () => {
+      checkAccess(ctx, 3);
+      if (Router.page() !== Page.Location) {
+        return { success: false, message: `Must be at the location screen, currently showing ${Router.page()}` };
+      }
+      const result = getInfilLocation(Player.location);
+      if (!result.success) {
+        return { success: false, message: result.message };
+      }
+      const location = result.location;
+      // Technically we don't need this check, but it's nicer to have it up-front.
+      // It is a duplicate, so watch out for divergence with the same code in Intro.tsx.
+      const startingSecurityLevel = location.infiltrationData.startingSecurityLevel;
+      const difficulty = calculateDifficulty(startingSecurityLevel);
+      if (difficulty >= MaxDifficultyForInfiltration) {
+        return {
+          success: false,
+          message: "This location is too secure for your current abilities. You cannot infiltrate it.",
+        };
+      }
+      Router.toPage(Page.Infiltration, { location, autoStart: true });
+      return { success: true };
+    },
+    getState: (__) => () => {
+      if (Router.page() !== Page.Infiltration) {
+        return null;
+      }
+      return getState();
+    },
+    pressKey: (ctx) => (_key) => {
+      const key = helpers.string(ctx, "key", _key);
+      const result = getInfilLocation(Player.location);
+      if (Router.page() !== Page.Infiltration || !result.success) {
+        // Perform a no-op if we're not infiltrating. The lack of delay should
+        // help clue people in who are debugging.
```

Comment:
Some script or console logging to these failure case(s) would help with debugging as well

Reply:
Done.